elasticsearch/src/test/scala/com/velocidi/apso/elasticsearch/ElasticsearchBulkInserterSpec.scala

```diff
@@ -85,22 +98,78 @@ class ElasticsearchBulkInserterSpec(implicit ee: ExecutionEnv) extends AkkaSpeci
       // insert a (valid) document
       bulkInserter ! Insert(Json.obj("name" := "test1"), msgIndex)
       bulkInserter ! Insert(Json.obj("name" := "test2"), msgIndex)
-      esClient.execute(searchQuery(msgIndex)).map(_.result.totalHits) must be_==(2).awaitFor(5.seconds).eventually(30, 2.seconds)
+      numberOfHits(msgIndex) must be_==(2).retry(eventuallyRetries = 30)
 
       // try to insert a (invalid) document; this one should stay on the buffer for retry later on...
       bulkInserter ! Insert(Json.obj("name" := "test3"), msgIndex)
       bulkInserter ! Insert(Json.obj("name" := "test4", "other" := "dynamic_field_value"), msgIndex)
-      esClient.execute(searchQuery(msgIndex)).map(_.result.totalHits) must not(be_==(4).awaitFor(5.seconds).eventually(5, 2.seconds))
-      esClient.execute(searchQuery(msgIndex)).map(_.result.totalHits) must be_==(3).awaitFor(5.seconds).eventually(5, 2.seconds)
+      numberOfHits(msgIndex) must not(be_==(4).retry())
+      numberOfHits(msgIndex) must be_==(3).retry()
 
       // now, change the mapping so that new fields are allowed...
       esClient.execute(putMapping(msgIndex) rawSource """{"dynamic":true,"properties":{"name":{"type":"text"}}}""") must
         beAnInstanceOf[RequestSuccess[_]].awaitFor(5.seconds)
       bulkInserter ! Insert(Json.obj("name" := "test5"), msgIndex)
-      esClient.execute(searchQuery(msgIndex)).map(_.result.totalHits) must
-        be_==(5).awaitFor(5.seconds).eventually(30, 2.seconds)
+      numberOfHits(msgIndex) must be_==(5).retry(eventuallyRetries = 30)
       esClient.execute(search(msgIndex) query matchQuery("other", "dynamic_field_value")).map(_.result.totalHits) must
-        be_==(1).awaitFor(5.seconds).eventually(30, 2.seconds)
+        be_==(1).retry(eventuallyRetries = 30)
+    }
+
+    "correctly send error back to client only when retries are exhausted" in {
+
+      // NOTICE: trying on a setting in which the retries are triggered by the buffer reaching its max capacity
```

Comment:
Maybe we could create separate tests for these isolated flows instead of separating them by a "NOTICE" comment?

Reply:
Sure, done in 229ff3c